src/Avalonia.Vulkan/VulkanControlBase.cs

```diff
@@ -15,13 +15,18 @@
 
 namespace Avalonia.Vulkan;
 
-public abstract class VulkanControlBase : Control
+public abstract class VulkanControlBase : Control, IDisposable
```

Comment:
I'm not quite sure that making a control to be disposable is a good idea. Any cleanup logic should happen in `OnDetachedFromVisualTree`

Reply:
OnDetachedFromVisualTree is also called when VulkanControlBase is inside a TabControl and when user changes Tabs. Also, if user decides to move the control from one parent to another, this should not dispose all Vulkan resources. Ok, the VulkanBitmapChain that is used by the VulkanControlBase could be disposed in OnDetachedFromVisualTree. But calling OnVulkanDeInit is too "heavy" - if user would be showing a complex 3D scene we should not dispose that because the control was hidden or moved. The user of VulkanControlBase should be aware that this control creates a lot of unmanaged resources and that it is his responsibility to clear that (for example, when the MainWindow is closed - as in the updated SilkNetDemo). When using only a single VulkanControlBase, then there is actually no need to clean anything because the OS will clean all resources; but when using multiple VulkanControlBase objects, user will need to manually dispose the resources when he knows they will no longer be used.

Reply:
VulkanControlBase is responsible for managing a swapchain and associated resources. When it's detached the swapchain becomes invalid and any associated resources like framebuffers should be disposed as well.


I'm currently working on something like SwapChainPanel and generic non-primary context creation APIs that would allow to manually manage resources that are shared between different vulkan-powered controls.

Reply:
Ok, I will remove the IDisposable from VulkanControlBase and will update the code so that in OnDetachedFromVisualTree the resources created by the VulkanControlBase will be disposed: VulkanBitmapChain and VulkanCommandBufferPool. But I would not call OnVulkanDeInit because OnDetachedFromVisualTree does not mean that all Vulkan resources should be disposed (actually I think OnVulkanDeInit should be removed because the VulkanControlBase cannot know when the Vulkan resources should be disposed). I would preserve the manual disposal of the Vulkan resource that are created in the VulkanDemoControl (they are disposed when the MainWindow is Unloaded). Is this ok?

Reply:
It still needs to inform the user code that swapchain images are no longer valid and that corresponding framebuffers should be removed.

Reply:
Yes ok, I will update the PR with that.

Reply:
https://github.com/AvaloniaUI/Avalonia/issues/9925 - see external memory interop spec.